templates/components/agents/typescript/form_filling/app/api/chat/workflow/formFilling.ts

```diff
@@ -0,0 +1,226 @@
+    });
+  }
+
+  private async prepareChatHistory(ctx: Context, ev: StartEvent<AgentInput>) {
+    const message = ev.data.input.message;
+
+    if (this.systemPrompt) {
+      this.memory.put({ role: "system", content: this.systemPrompt });
+    }
+    this.memory.put({ role: "user", content: message });
+
+    return new InputEvent({ input: this.memory.getMessages() });
+  }
+
+  private async handleLLMInput(ctx: Context, ev: InputEvent) {
+    const chatHistory = ev.data.input;
+
+    const tools = [this.extractorTool, this.fillMissingCellsTool];
+    if (this.queryEngineTool) {
+      tools.push(this.queryEngineTool);
+    }
+
+    const toolCallResponse = await chatWithTools(this.llm, tools, chatHistory);
+
+    if (!toolCallResponse.isCallingTool()) {
+      return new StopEvent({ result: toolCallResponse.responseGenerator });
+    }
+
+    if (toolCallResponse.isCallingDifferentTools()) {
+      this.memory.put({
+        role: "user",
+        content:
+          "Calling different tool is not allowed. Please only call one tool at a time.",
+      });
+      return new InputEvent({ input: this.memory.getMessages() });
+    }
+
+    // Put the LLM tool call message into the memory
+    // And trigger the next step according to the tool call
+    if (toolCallResponse.toolCallMessage) {
+      this.memory.put(toolCallResponse.toolCallMessage);
+    }
+    switch (toolCallResponse.toolName()) {
+      case this.extractorTool.metadata.name:
+        return new ExtractMissingCellsEvent({
+          toolCalls: toolCallResponse.toolCalls,
+        });
+      case this.fillMissingCellsTool.metadata.name:
+        return new FillMissingCellsEvent({
+          toolCalls: toolCallResponse.toolCalls,
+        });
+      case this.queryEngineTool?.metadata.name:
+        return new FindAnswersEvent({
+          toolCalls: toolCallResponse.toolCalls,
+        });
+    }
+  }
+
+  private async handleExtractMissingCells(
+    ctx: Context,
+    ev: ExtractMissingCellsEvent,
+  ) {
+    ctx.writeEventToStream(
+      new AgentRunEvent({
+        name: "CSVExtractor",
+        text: "Extracting missing cells",
+        type: "text",
+      }),
+    );
+    const { toolCalls } = ev.data;
+    const toolMsgs = await callTools(
+      toolCalls,
+      [this.extractorTool],
+      ctx,
+      "CSVExtractor",
+    );
+    for (const toolMsg of toolMsgs) {
+      this.memory.put(toolMsg);
+    }
+    return new InputEvent({ input: this.memory.getMessages() });
+  }
+
+  private async handleFindAnswers(ctx: Context, ev: FindAnswersEvent) {
+    const { toolCalls } = ev.data;
+    if (!this.queryEngineTool) {
+      throw new Error("Query engine tool is not available");
+    }
+    ctx.writeEventToStream(
+      new AgentRunEvent({
+        name: "Researcher",
+        text: "Finding answers",
+        type: "text",
+      }),
+    );
+    const toolMsgs = await callTools(
+      toolCalls,
+      [this.queryEngineTool],
+      ctx,
+      "Researcher",
+    );
```

Comment:
easier to read (and later to refactor if we use a param object)
```suggestion
    const toolMsgs = await callTools({
     ctx,
     name: "Researcher",
     tools: [this.queryEngineTool],
     toolCalls,
    });
```